### PR TITLE
fix(core): Account for addEventListener to be passed a Window or Document

### PR DESCRIPTION
**Pull request by iteriani**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [xx] No

### CODE REVIEW

File: packages/core/src/event_delegation_utils.ts [hunk 1]
```diff
@@ -93,15 +93,31 @@ export class GlobalEventDelegation implements OnDestroy {
...
+    element: HTMLElement | Window | Document,
+    eventType: string,
+    handler: Function,
+  ): Function {
+    if (element instanceof HTMLElement) {
+      this.eventContractDetails.instance!.addEvent(eventType);
+      getActionCache(element)[eventType] = '';
+      sharedStashFunction(element, eventType, handler);
+    } else {
+      element.addEventListener(eventType, handler as EventListener);
```
Comment (AndrewKushnir) [inline packages/core/src/event_delegation_utils.ts:106]: nit: it'd be great to add a quick comment here in the code to explain which objects does this code handle.

Comment (AndrewKushnir) [reply packages/core/src/event_delegation_utils.ts:106]: (similarly in the code below)

File: packages/core/test/event_dispatch/event_dispatch_spec.ts [hunk 2]
```diff
@@ -6,7 +6,14 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import {Component, Renderer2, inject, ɵprovideGlobalEventDelegation} from '@angular/core';
+import {
+  Component,
+  ElementRef,
+  Renderer2,
+  ViewChild,
```
Comment (AndrewKushnir) [inline packages/core/test/event_dispatch/event_dispatch_spec.ts:13]: It looks like we are adding extra symbols here (`ElementRef` and `ViewChild`) that we don't use.

Unrelated: would it be possible to put together a test for the updated logic?

File: packages/core/src/event_delegation_utils.ts [hunk 3]
```diff
@@ -93,15 +93,31 @@ export class GlobalEventDelegation implements OnDestroy {
...
-    this.eventContractDetails.instance!.addEvent(eventName);
-    sharedStashFunction(element, eventName, handler);
-    getActionCache(element)[eventName] = '';
-    return () => this.removeEventListener(element, eventName, handler);
+  addEventListener(
+    element: HTMLElement | Window | Document,
+    eventType: string,
+    handler: Function,
+  ): Function {
+    if (element instanceof HTMLElement) {
```
Comment (crisbeto) [inline packages/core/src/event_delegation_utils.ts:101]: I don't know if this code can run on the server, but `instanceof HTMLElement` can break during SSR. On Material we use something like `element.nodeType === 1` as a safer way to check.

File: packages/core/src/event_delegation_utils.ts (hunk 1 above)
Comment (iteriani) [reply packages/core/src/event_delegation_utils.ts:104]: done

File: packages/core/src/event_delegation_utils.ts (hunk 3 above)
Comment (iteriani) [reply packages/core/src/event_delegation_utils.ts]: done

File: packages/core/test/event_dispatch/event_dispatch_spec.ts (hunk 2 above)
Comment (iteriani) [reply packages/core/test/event_dispatch/event_dispatch_spec.ts]: done

**Review by AndrewKushnir — APPROVED** [review]
👍

**Review by thePunderWoman — APPROVED** [review]
LGTM

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 3439cc2049a92c9a68fedd4354940429b2aa3ff1.

The changes were merged into the following branches: main, 18.1.x, 18.2.x

**Pull request closed without merging**